From 86b7236edcb4e1719a0da7273128286b3fda20eb Mon Sep 17 00:00:00 2001
From: Ahmad Sharif <asharif@chromium.org>
Date: Tue, 24 Apr 2012 11:19:19 -0700
Subject: [PATCH] Added TLS size to stack size if it's lower than 16 *
 TLS_SIZE.

Under normal circumstances, the TLS size is substracted from the stack size
before allocaton. This can cause the application to crash if the TLS size is
close to the stack size.
---
 nptl/allocatestack.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/nptl/allocatestack.c b/nptl/allocatestack.c
index 79c4531..7864c8c 100644
--- a/nptl/allocatestack.c
+++ b/nptl/allocatestack.c
@@ -463,6 +463,9 @@ allocate_stack (const struct pthread_attr *attr, struct pthread **pdp,
 	size += pagesize_m1 + 1;
 #endif
 
+      if (size < 16 * __static_tls_size)
+        size = roundup (size + __static_tls_size, pagesize_m1 + 1);
+
       /* Adjust the stack size for alignment.  */
       size &= ~__static_tls_align_m1;
       assert (size != 0);
-- 
1.7.7.3

